### ISSUE TITLE
how to generate centers in MMC

### ISSUE THREAD

**Issue by moicesong** [P2333/Max-Mahalanobis-Training#1]
#Load centers in MMC
kernel_dict = loadmat('kernel_paras/meanvar1_featuredim'+str(feature_dim)+'_class'+str(num_class)+'.mat')
mean_logits_np = kernel_dict['mean_logits'] #num_class X num_dense
mean_logits = FLAGS.mean_var * tf.constant(mean_logits_np,dtype=tf.float32)

can give some code implementation?

**Comment by P2333** [comment]
The crafting code is provided in [kernel_paras/craft_MMcenters.m](https://github.com/P2333/Max-Mahalanobis-Training/blob/master/kernel_paras/craft_MMcenters.m) using MATLAB.

**Comment by moicesong** [comment]
thinks

**Comment by kirk86** [comment]
@P2333 Hi, if you don't mind me asking an equivalent question.

I see in the train file the following
```
kernel_dict = loadmat('kernel_paras/meanvar1_featuredim'+str(feature_dim)+'_class'+str(num_class)+'.mat')
mean_logits_np = kernel_dict['mean_logits'] #num_class X num_dense
mean_logits = FLAGS.mean_var * tf.constant(mean_logits_np,dtype=tf.float32)  <--- is this necessary?
```
I was wondering if the multiplication of the variance with the means of logits is necessary since its already done once in the matlab script?

I've translated the matlab script into python and by looking the script, correct me if I'm wrong but it seems that `mean_logits` is supposed to be lower triangular matrix produced by the matlab script?

Usually when I run it I get all zeros except from 2-3 values. I don't understand how you're supposed to compute the means of the logits without having the logits in the first place? The script doesn't use anywhere the logits, is this supposed to randomly assign means for the logits?

Thanks!

**Comment by P2333** [comment]
@kirk86 Thank you for the questions.

As to the variance flag, we craft the mean_logit with variance=1 (meanvar1) from the Matlab script and tune the variance by FLAGS.mean_var in python. This is a little bit more flexible than crafting .mat files for different variance values.

The mean_logits, i.e., the MM centers are data-agnostic and preset before training. They only depend on the dimension of feature space `d` and the number of classes `L`. For example, in the case of `L=3` and `d=2`, we want to separate three points on a 2-d plane, such that the minimal distance between any two points is maximal, given that the norm of each point is constrained by `C`. Then we can figure out that we should set these three points on the vertexes of an equilateral triangle (I provide a demo in readme).

In high-dimension space, e.g., for a Resnet trained on CIFAR-10, we may have `L=10` and `d=256`. Since any `L` points will locate in a `L-1`-d subspace, we craft a special instantiation of the MM centers in our Matlab script. After any rotation transformation, the MM centers are still valid. While during training, we expect the data feature points to gather around the preset MM centers.

**Comment by kirk86** [comment]
@P2333 Thanks for your reply.

If you don't mind me asking a follow up question. 

> In high-dimension space, e.g., for a Resnet trained on CIFAR-10, we may have `L=10` and `d=256`

That's what I initially understood that `d` always corresponds to the dimensionality of data `X = num_instances x dim_featuers`. For instance if 
I have `y = f(x)` where `y = batch_size x num_classes` is the logits then `d` should be set to match the dimension of the of the data `X = num_instances x dim_features`, such that `d = dim_featues`? Not the output from the penultimate layer of `f`? 

For instance, suppose we have a simple example where my data `X = num_instances x 2`, so its a 2-d dataset, and we have a network `y = f(x)` where `y = batch_size x 3` is the logits and that `y_pen = f(x)` is the output of the penultimate layer of dim `batch_size x 50`, then according to your explanation `L=3` and `d = 2` and not `d=50`, correct?

> we craft a special instantiation of the MM centers in our Matlab script.

Is this different from the already provided script capabilities?
I think there might be an error in the matlab script due to current behaviour but could be intentional? 
If I run your matlab script with `L=3` and `D=2` it returns an array of `3x3` while it should return `3x2`.
The error is in line `9` where when `k=3` you are calling `mean_logits[3, 3]` but column `3` doesn't exist since you've created `mean_logits = zeros(L, D)`, but matlab adds that additional column silently while in other languages you'll get an error for referencing a column that doesn't exist.

![image](https://user-images.githubusercontent.com/2902390/85535647-4adac180-b60a-11ea-891c-63c571268fe9.png)


Here are some more interesting cases. If I call your matlab script with `D = 1` and `D = -1`, I'll get the same exact results as with `D = 2`. That doesn't seem correct?

![image](https://user-images.githubusercontent.com/2902390/85536823-58447b80-b60b-11ea-8fc3-5698e7c6813c.png)

**Comment by P2333** [comment]
@kirk86 Yes, `d` refers to the dimension of feature space (the penultimate layer), rather than the input space.

As to the Matlab script, you can simply ignore or squeeze the zero elements to the dimension you want. For the cases of `d = 1` and `d = -1`, the reason is the same as you mentioned that Matlab will add the additional column silently, so the Matlab script will return same results when `L > d`.

**Comment by kirk86** [comment]
@P2333 Thanks a lot, appreciate it!

One last question, so the MM centres don't change during training? My understanding is that they are computed only once before training and for any data point `x in X` they remain the same, is that correct?

**Comment by P2333** [comment]
@kirk86 Yes, the MM centers are fixed during training.

In the Appendix B.3 of [our ICLR 2020 paper](https://openreview.net/pdf?id=Byg9A24tvB), we provide more flexible variants, e.g., regarding the MM centers as a prior rather than a hard constraint.

**Comment by kirk86** [comment]
@P2333 Hi, sorry to bother you again.
I was wondering if you have any insights regarding matching the results on Table 1 on page 8.
I've trained a wideresnet using MMLDA + dot_loss, and get 95.18 accuracy on cifar10 clean test set.
Then use PGD un-targeted  with epsilon 8/255 and 20 iterations, but accuracy on the adversarial cifar10 test set is 93.24.
I'm not approximate the values in Table 1, do you have any thoughts what I might be missing?

Thanks!

**Comment by P2333** [comment]
@kirk86 To evade our method, you need to apply adaptive attacks. Note that our evaluation is based on cleverhans: 2.1.0. To perform adaptive attack, please manually modify the function `model_loss` of the file `utils_tf.py` by substituting the softmax cross-entropy loss with other adaptive objectives, e.g., `out=-tf.reduce_sum(logits * y, axis=-1)`.

**Comment by kirk86** [comment]
@P2333 Thank you!
I think my mistake was that I had forgotten the `softmax` loss in my function when computing PGD with l-infinity. 
I wasn't using cleverhans, but now I've changed `cross-entropy` with `out=torch.mean(logits * y)` and now the results on corrupted test set with adv. noise is 45.71 for `MMLDA + dot_loss`.

But I'm still a bit confused regarding Table 1 on page 8, about the difference between `MMLDA + dot_loss` vs `MMC`, aren't they the same thing?

**Comment by P2333** [comment]
@kirk86 `MMLDA + dot_loss` is exactly `MMC`. 
This name schedule is a "historical legacy" since the MMC loss is based on our MMLDA method without the softmax normalization. Thus in our code, we do not explicitly define a flag for MMC, and regard MMC as a modified version of MMLDA with dot loss.

**Comment by kirk86** [comment]
@P2333 Great, thank you so much. Appreciate it!
